From 28bbf0f62f74792c8ef3762ba8a904ba5d780d52 Mon Sep 17 00:00:00 2001
From: Hien Huynh <hien.huynh.px@renesas.com>
Date: Thu, 30 May 2019 15:02:21 +0700
Subject: [PATCH 076/458] sched/fair: Don't move tasks to lower capacity cpus
 unless necessary

When lower capacity CPUs are load balancing and considering to pull
something from a higher capacity group, we should not pull tasks
from a cpu with only one task running as this is guaranteed to impede
progress for that task. If there is more than one task running, load
balance in the higher capacity group would have already made any
possible moves to resolve imbalance and we should make better use of
system compute capacity by moving a task if we still have more than one
running.

Cherry-picked from https://patchwork.kernel.org/patch/10506527/

cc: Ingo Molnar <mingo@redhat.com>
cc: Peter Zijlstra <peterz@infradead.org>

Signed-off-by: Chris Redpath <chris.redpath@arm.com>
Signed-off-by: Morten Rasmussen <morten.rasmussen@arm.com>
Signed-off-by: Gaku Inami <gaku.inami.xh@renesas.com>
Signed-off-by: Hien Huynh <hien.huynh.px@renesas.com>
---
 kernel/sched/fair.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index e0e17cd..35c2b7e 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -8646,6 +8646,17 @@ static struct rq *find_busiest_queue(struct lb_env *env,
 
 		capacity = capacity_of(i);
 
+		/*
+		 * For ASYM_CPUCAPACITY domains, don't pick a cpu that could
+		 * eventually lead to active_balancing high->low capacity.
+		 * Higher per-cpu capacity is considered better than balancing
+		 * average load.
+		 */
+		if (env->sd->flags & SD_ASYM_CPUCAPACITY &&
+				capacity_of(env->dst_cpu) < capacity &&
+				rq->nr_running == 1)
+			continue;
+
 		wl = weighted_cpuload(rq);
 
 		/*
-- 
2.7.4

